blockstore/splitstore/markset_badger.go

```diff
@@ -188,6 +136,140 @@ func (s *BadgerMarkSet) Has(c cid.Cid) (bool, error) {
 	}
 }
 
+func (s *BadgerMarkSet) Visit(c cid.Cid) (bool, error) {
+	key := c.Hash()
+	pendKey := string(key)
+
+	checkPending := func() (bool, error) {
+		if s.pend == nil {
+			return false, errMarkSetClosed
+		}
+
+		if _, ok := s.pend[pendKey]; ok {
+			return false, nil
+		}
+
+		for _, wr := range s.writing {
+			if _, ok := wr[pendKey]; ok {
+				return false, nil
+			}
+		}
+
+		return true, nil
+	}
+
+	s.mx.RLock()
+
+	visit, err := checkPending()
+	if !visit {
+		s.mx.RUnlock()
+		return visit, err
+	}
+
+	err = s.db.View(func(txn *badger.Txn) error {
+		_, err := txn.Get(key)
+		return err
+	})
+
+	switch err {
+	case nil:
+		s.mx.RUnlock()
+		return false, nil
+
+	case badger.ErrKeyNotFound:
+		// need to upgrade the lock to exclusive in order to write; take the version count to see
+		// if there was another write while we were upgrading
+		version := s.version
+		s.mx.RUnlock()
+
+		s.mx.Lock()
+
+		// we have to do the check dance again
+		visit, err := checkPending()
+		if !visit {
+			s.mx.Unlock()
+			return visit, err
+		}
+
+		if version != s.version {
+			// something was written to the db, we need to check it
+			err = s.db.View(func(txn *badger.Txn) error {
+				_, err := txn.Get(key)
+				return err
+			})
+
+			switch err {
+			case nil:
+				s.mx.Unlock()
+				return false, nil
+
+			case badger.ErrKeyNotFound:
+
+			default:
+				s.mx.Unlock()
+				return false, xerrors.Errorf("error checking badger markset: %w", err)
+			}
+		}
+
+		s.pend[pendKey] = struct{}{}
+		if len(s.pend) < badgerMarkSetBatchSize {
+			s.mx.Unlock()
+			return true, nil
+		}
+
+		if err := s.putPending(); err != nil {
+			return false, err
+		}
+
+		return true, nil
+
+	default:
+		s.mx.RUnlock()
+		return false, xerrors.Errorf("error checking badger markset: %w", err)
+	}
+}
+
+// writer holds the lock
+func (s *BadgerMarkSet) putPending() error {
+	pend := s.pend
+	seqno := s.seqno
+	s.seqno++
+	s.writing[seqno] = pend
+	s.pend = make(map[string]struct{})
+	s.writers++
```

Comment:
Isn't this redundant with `len(s.writing)`?

Reply:
yes it is.

Reply:
removed.